sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.c

```diff
@@ -0,0 +1,195 @@
+ *
+ * @file
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ */
+
+#include <stdbool.h>
+
+#include "luid.h"
+#include "net/gnrc/netif/internal.h"
+
+#include "_nib-arsm.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+
+#if GNRC_IPV6_NIB_CONF_6LN || GNRC_IPV6_NIB_CONF_SLAAC
+void _auto_configure_addr(gnrc_netif_t *netif, const ipv6_addr_t *pfx,
+                          uint8_t pfx_len)
+{
+    ipv6_addr_t addr = IPV6_ADDR_UNSPECIFIED;
+    int idx;
+    uint8_t flags = GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_TENTATIVE;
+
+    DEBUG("nib: add address based on %s/%u automatically to interface %u\n",
+          ipv6_addr_to_str(addr_str, pfx, sizeof(addr_str)),
+          pfx_len, netif->pid);
+#if GNRC_IPV6_NIB_CONF_6LN
+    bool new_address = false;
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+    gnrc_netif_ipv6_get_iid(netif, (eui64_t *)&addr.u64[1]);
+    ipv6_addr_init_prefix(&addr, pfx, pfx_len);
+    if ((idx = gnrc_netif_ipv6_addr_idx(netif, &addr)) < 0) {
+        if ((idx = gnrc_netif_ipv6_addr_add_internal(netif, &addr, pfx_len,
+                                                     flags)) < 0) {
+            DEBUG("nib: Can't add link-local address on interface %u\n",
+                  netif->pid);
+            return;
+        }
+#if GNRC_IPV6_NIB_CONF_6LN
+        new_address = true;
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+    }
+
+#if GNRC_IPV6_NIB_CONF_6LN
+    /* mark link-local addresses as valid on 6LN */
+    if (gnrc_netif_is_6ln(netif) && ipv6_addr_is_link_local(pfx)) {
+        /* don't do this beforehand or risk a deadlock:
+         *  * gnrc_netif_ipv6_addr_add_internal() adds VALID (i.e. manually configured
+         *    addresses to the prefix list locking the NIB's mutex which is already
+         *    locked here) */
+        netif->ipv6.addrs_flags[idx] &= ~GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_MASK;
+        netif->ipv6.addrs_flags[idx] |= GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_VALID;
+    }
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+#if GNRC_IPV6_NIB_CONF_6LN
+    if (new_address && gnrc_netif_is_6ln(netif) &&
+        !gnrc_netif_is_6lbr(netif)) {
+        _handle_rereg_address(&netif->ipv6.addrs[idx]);
+    }
+#else   /* GNRC_IPV6_NIB_CONF_6LN */
+    (void)idx;
+#endif  /* GNRC_IPV6_NIB_CONF_6LN */
+}
+#endif  /* GNRC_IPV6_NIB_CONF_6LN || GNRC_IPV6_NIB_CONF_SLAAC */
+
+#if GNRC_IPV6_NIB_CONF_SLAAC
+static bool _try_l2addr_reconfiguration(gnrc_netif_t *netif)
+{
+    uint8_t hwaddr[GNRC_NETIF_L2ADDR_MAXLEN];
+    eui64_t orig_iid;
+    bool remove_old = false;
+    uint16_t hwaddr_len;
+
+    if (gnrc_netif_ipv6_get_iid(netif, &orig_iid) == 0) {
+        remove_old = true;
+    }
+    /* seize netif to netif thread, since gnrc_netapi_get/gnrc_netapi_set are
+     * synchronous, this is safe */
+    gnrc_netif_release(netif);
+    if (gnrc_netapi_get(netif->pid, NETOPT_SRC_LEN, 0, &hwaddr_len,
+                        sizeof(hwaddr_len)) < 0) {
+        goto error_exit;
```

Comment:
Usually I'm not a fan of `goto` as well. I can do

```C
do {
    /* blafoo */
    if (error_condition) {
        break;
    }
    return true;
} while(0);
gnrc_netif_acquire(netif);
return false;
```

as an alternative. But people weren't a fan of that either. I try to reduce code duplication here. Also, to answer your question for current master:

```
$ git grep "^\s\+\<goto\>" -- sys/ examples/ | wc -l
18
```

In all cases it is used in similar fashion as it is here.

Reply:
the latter sounds like: "all the others are doing it why do you blame me" 😀 

Anyway, as far as I can see its "only" needed because of `gnrc_netif_acquire(netif);`, can't this be moved to the calling function(s) instead so you get rid of `gnrc_netif_acquire` and `gnrc_netif_release` in all private/internal functions used here? Or at least in such away that acquire and release are handled by the calling function (if if that is private/internal), like [here](https://github.com/RIOT-OS/RIOT/blob/cea0d132bfcd1b8c9ca9a39426538ab440d6b4c0/sys/net/gnrc/network_layer/ipv6/nib/_nib-router.c#L35) for instance? Because then you could remove the goto and just return true or false instead..